File: src/services/spotifyService.ts
// src/services/spotifyService.ts
import * as SecureStore from "expo-secure-store";
import {
  SpotifyTrack,
  SpotifyUser,
  SpotifyPlaylist,
  VibeMode,
} from "../types/music";
import { SPOTIFY_CONFIG } from "../config/spotify";

const SPOTIFY_BASE_URL = "https://api.spotify.com/v1";
const SPOTIFY_ACCOUNTS_URL = "https://accounts.spotify.com/api/token";

class SpotifyService {
  /** --------------------------
   * TOKEN STORAGE HELPERS
   * ------------------------- */
  public async getAccessToken(): Promise<string | null> {
    const token = await SecureStore.getItemAsync("spotify_access_token").catch(() => null);
    console.log("🔑 Retrieved token:", token);
    return token;
  }

  public async setAccessToken(token: string): Promise<void> {
    await SecureStore.setItemAsync("spotify_access_token", token).catch(console.error);
    console.log("🔑 Set Access token:", token);
  }

  public async getRefreshToken(): Promise<string | null> {
    return SecureStore.getItemAsync("spotify_refresh_token").catch(() => null);
  }

  public async setRefreshToken(token: string): Promise<void> {
    await SecureStore.setItemAsync("spotify_refresh_token", token).catch(console.error);
  }

  async saveTokens(accessToken: string, refreshToken: string): Promise<void> {
    console.log("💾 Saving tokens:", { accessToken, refreshToken });
    await this.setAccessToken(accessToken);
    if (refreshToken) await this.setRefreshToken(refreshToken);
    console.log("🔑 Refresh Token:", refreshToken);
  }

  async clearTokens(): Promise<void> {
    await SecureStore.deleteItemAsync("spotify_access_token").catch(console.error);
    await SecureStore.deleteItemAsync("spotify_refresh_token").catch(console.error);
  }

  /** --------------------------
   * GENERIC API CALL
   * ------------------------- */
  private async makeRequest<T>(endpoint: string, options: RequestInit = {}): Promise<T> {
    let accessToken = await this.getAccessToken();
    if (!accessToken) throw new Error("No access token available");

    const url = endpoint.startsWith("http") ? endpoint : `${SPOTIFY_BASE_URL}${endpoint}`;

    const response = await fetch(url, {
      ...options,
      headers: {
        Authorization: `Bearer ${accessToken}`,
        "Content-Type": "application/json",
        ...options.headers,
      },
    });

    // Handle expired token
    if (response.status === 401) {
      const refreshed = await this.refreshAccessToken();
      if (refreshed) return this.makeRequest<T>(endpoint, options);
      throw new Error("Authentication failed");
    }

    // Handle other API errors
    if (!response.ok) {
      const errorText = await response.text();
      console.error(`Spotify API error ${response.status}:`, errorText);
      throw new Error(`Spotify API error: ${response.status}`);
    }

    // Handle empty responses (like 204 No Content for DELETE requests)
    const contentType = response.headers.get('content-type');
    if (!contentType || !contentType.includes('application/json')) {
      return {} as T;
    }

    // Check if response has content before trying to parse JSON
    const text = await response.text();
    if (!text.trim()) {
      return {} as T;
    }

    return JSON.parse(text);
  }

  /** --------------------------
   * TOKEN REFRESH (PKCE SAFE)
   * ------------------------- */
  async refreshAccessToken(): Promise<boolean> {
    const refreshToken = await this.getRefreshToken();
    if (!refreshToken) {
      console.error("⚠️ No refresh token found");
      return false;
    }

    try {
      const response = await fetch(SPOTIFY_ACCOUNTS_URL, {
        method: "POST",
        headers: {
          "Content-Type": "application/x-www-form-urlencoded",
        },
        body: new URLSearchParams({
          grant_type: "refresh_token", // ✅ required for PKCE
          refresh_token: refreshToken,
          client_id: SPOTIFY_CONFIG.CLIENT_ID, // ✅ PKCE needs client_id, not secret
        }).toString(),
      });

      const data = await response.json();

      if (!response.ok) {
        console.error("❌ Failed to refresh token:", data);
        return false;
      }

      if (data.access_token) {
        await this.setAccessToken(data.access_token);

        // Some refresh responses also return a new refresh token
        if (data.refresh_token) {
          await this.setRefreshToken(data.refresh_token);
        }

        console.log("✅ Spotify token refreshed successfully");
        return true;
      }

      console.error("❌ Failed to refresh token, no access token returned:", data);
      return false;
    } catch (err) {
      console.error("Error refreshing token:", err);
      return false;
    }
  }


  /** --------------------------
   * SPOTIFY API METHODS
   * ------------------------- */
  async getCurrentUser(): Promise<SpotifyUser> {
    return this.makeRequest<SpotifyUser>("/me");
  }

  async getUserTopTracks(
    timeRange: "short_term" | "medium_term" | "long_term" = "medium_term"
  ): Promise<SpotifyTrack[]> {
    const endpoint = `/me/top/tracks?time_range=${timeRange}&limit=50`;
    console.log("🎯 Fetching user top tracks from endpoint:", endpoint);

    try {
      const res = await this.makeRequest<{ items: SpotifyTrack[] }>(endpoint);
      console.log("✅ Top tracks response items:", res.items?.length);
      return res.items ?? [];
    } catch (err: any) {
      console.error("❌ Failed to fetch top tracks:", err);
      // Return empty array to prevent breaking recommendations
      return [];
    }
  }

  async getUserTopArtists(
    timeRange: "short_term" | "medium_term" | "long_term" = "medium_term"
  ) {
    const res = await this.makeRequest<{ items: any[] }>(
      `/me/top/artists?time_range=${timeRange}&limit=50`
    );
    return res.items;
  }

  async getRecommendations(params: { query?: string; limit?: number }): Promise<SpotifyTrack[]> {
    const query = params.query || "pop"; // fallback
    const limit = params.limit || 50;

    console.log("🎯 Searching Spotify for:", query);

    try {
      const tracks = await this.searchTracks(query, limit);
      console.log(`✅ Found ${tracks.length} tracks for query "${query}"`);
      return tracks;
    } catch (err: any) {
      console.error("❌ Spotify search failed:", err.message);
      return [];
    }
  }

  async getRecommendationsBasedOnTracks(userTopTracks: SpotifyTrack[]): Promise<SpotifyTrack[]> {
    console.log("🎯 Finding songs similar to your top tracks using metadata analysis");
    
    const allTracks: SpotifyTrack[] = [];
    
    // Extract characteristics from your top tracks (no audio features needed)
    const characteristics = this.analyzeTrackCharacteristics(userTopTracks);
    console.log("🎵 Your music characteristics:", characteristics);
    
    // Search for similar artists from your top tracks
    const topArtists = [...new Set(userTopTracks.slice(0, 20).map(t => t.artists[0]?.name).filter(Boolean))];
    
    for (const artist of topArtists.slice(0, 10)) { // Top 10 artists
      try {
        const tracks = await this.searchTracks(`artist:"${artist}"`, 30);
        allTracks.push(...tracks);
        console.log(`Found ${tracks.length} tracks by "${artist}"`);
      } catch (err) {
        console.warn(`Failed to search for artist "${artist}":`, err);
      }
    }
    
    // Search for similar genres based on your top tracks
    const genreTerms = this.extractGenreTerms(userTopTracks);
    for (const genre of genreTerms.slice(0, 6)) {
      try {
        const tracks = await this.searchTracks(genre, 25);
        allTracks.push(...tracks);
        console.log(`Found ${tracks.length} tracks for genre "${genre}"`);
      } catch (err) {
        console.warn(`Failed to search for genre "${genre}":`, err);
      }
    }
    
    // Search for tracks with similar popularity ranges
    const popularityRanges = this.getPopularityRanges(userTopTracks);
    for (const range of popularityRanges) {
      try {
        const tracks = await this.searchTracks(`year:${range.year}`, 20);
        allTracks.push(...tracks);
        console.log(`Found ${tracks.length} tracks from ${range.year}`);
      } catch (err) {
        console.warn(`Failed to search for year ${range.year}:`, err);
      }
    }
    
    // Search for tracks with similar popularity levels
    const avgPopularity = userTopTracks.reduce((sum, track) => sum + track.popularity, 0) / userTopTracks.length;
    if (avgPopularity > 70) {
      // High popularity - search for popular tracks
      try {
        const tracks = await this.searchTracks("popular", 20);
        allTracks.push(...tracks);
        console.log(`Found ${tracks.length} popular tracks`);
      } catch (err) {
        console.warn(`Failed to search for popular tracks:`, err);
      }
    } else if (avgPopularity < 30) {
      // Low popularity - search for indie/underground tracks
      try {
        const tracks = await this.searchTracks("indie", 20);
        allTracks.push(...tracks);
        console.log(`Found ${tracks.length} indie tracks`);
      } catch (err) {
        console.warn(`Failed to search for indie tracks:`, err);
      }
    }
    
    // Remove duplicates and return up to 300 tracks
    const uniqueTracks = allTracks.filter((track, index, self) => 
      index === self.findIndex(t => t.id === track.id)
    );
    
    console.log(`Found ${uniqueTracks.length} unique similar tracks`);
    return uniqueTracks.slice(0, 300);
  }

  async getRecommendationsForVibeMode(
    vibeMode: VibeMode | null,
    userTopTracks: SpotifyTrack[]
  ): Promise<SpotifyTrack[]> {
    if (!vibeMode) {
      console.log("⚠️ No vibe mode provided — returning user's top tracks");
      return userTopTracks.slice(0, 50);
    }

    console.log("🎯 Getting tracks for vibe mode:", vibeMode.name);
    
    // Start with empty array - only vibe-specific tracks
    const allTracks: SpotifyTrack[] = [];
    
    // Add vibe-specific searches
    const vibeSearches = [
      `${vibeMode.name} music`,
      `popular ${vibeMode.name}`,
      `${vibeMode.name} hits`,
      `trending ${vibeMode.name}`,
      `best ${vibeMode.name} songs`,
      `${vibeMode.name} 2024`,
      `${vibeMode.name} 2023`,
      `top ${vibeMode.name} artists`
    ];
    
    // Search for vibe-specific tracks
    for (const searchTerm of vibeSearches) {
      try {
        const tracks = await this.searchTracks(searchTerm, 30);
        allTracks.push(...tracks);
        console.log(`Found ${tracks.length} tracks for "${searchTerm}"`);
      } catch (err) {
        console.warn(`Failed to search for "${searchTerm}":`, err);
      }
    }
    
    // Add some popular tracks by searching for well-known artists
    const popularArtists = [
      "Drake", "Taylor Swift", "The Weeknd", "Billie Eilish", "Ariana Grande",
      "Ed Sheeran", "Post Malone", "Dua Lipa", "Olivia Rodrigo", "Harry Styles",
      "Bad Bunny", "Travis Scott", "Kendrick Lamar", "J. Cole", "Future"
    ];
    
    for (const artist of popularArtists) {
      try {
        const tracks = await this.searchTracks(`artist:${artist}`, 25);
        allTracks.push(...tracks);
      } catch (err) {
        console.warn(`Failed to search for artist "${artist}":`, err);
      }
    }
    
    // Remove duplicates and return up to 300 tracks
    const uniqueTracks = allTracks.filter((track, index, self) => 
      index === self.findIndex(t => t.id === track.id)
    );
    
    console.log(`Found ${uniqueTracks.length} unique tracks for vibe mode`);
    return uniqueTracks.slice(0, 300);
  }

  async getUserPlaylists(): Promise<SpotifyPlaylist[]> {
    try {
      const res = await this.makeRequest<{ items: SpotifyPlaylist[] }>("/me/playlists?limit=50");
      return res.items || [];
    } catch (error) {
      console.error("Error fetching user playlists:", error);
      return [];
    }
  }

  async getPlaylistTracks(playlistId: string): Promise<SpotifyTrack[]> {
    try {
      const res = await this.makeRequest<{ items: { track: SpotifyTrack }[] }>(`/playlists/${playlistId}/tracks?limit=100`);
      // Filter out null tracks (some playlists have null tracks)
      return res.items?.map(item => item.track).filter(track => track && track.id) || [];
    } catch (error) {
      console.error("Error fetching playlist tracks:", error);
      return [];
    }
  }

  async createPlaylist(name: string, description?: string): Promise<SpotifyPlaylist> {
    const user = await this.getCurrentUser();
    return this.makeRequest<SpotifyPlaylist>(`/users/${user.id}/playlists`, {
      method: "POST",
      body: JSON.stringify({ name, description: description || "", public: false }),
    });
  }

  async addTracksToPlaylist(playlistId: string, trackUris: string[]): Promise<void> {
    await this.makeRequest(`/playlists/${playlistId}/tracks`, {
      method: "POST",
      body: JSON.stringify({ uris: trackUris }),
    });
  }

  async removeTracksFromPlaylist(playlistId: string, trackUris: string[]): Promise<void> {
    await this.makeRequest(`/playlists/${playlistId}/tracks`, {
      method: "DELETE",
      body: JSON.stringify({ uris: trackUris }),
    });
  }

  async addTrackToLikedSongs(trackId: string): Promise<void> {
    await this.makeRequest(`/me/tracks`, {
      method: "PUT",
      body: JSON.stringify({ ids: [trackId] }),
    });
  }

  async removeTrackFromLikedSongs(trackId: string): Promise<void> {
    await this.makeRequest(`/me/tracks`, {
      method: "DELETE",
      body: JSON.stringify({ ids: [trackId] }),
    });
  }

  async searchTracks(query: string, limit = 20): Promise<SpotifyTrack[]> {
    const res = await this.makeRequest<{ tracks: { items: SpotifyTrack[] } }>(
      `/search?q=${encodeURIComponent(query)}&type=track&limit=${limit}`
    );
    return res.tracks.items;
  }

  private analyzeTrackCharacteristics(tracks: SpotifyTrack[]): any {
    const avgPopularity = tracks.reduce((sum, track) => sum + track.popularity, 0) / tracks.length;
    
    return {
      avgPopularity,
      genres: this.extractGenreTerms(tracks),
      years: this.getPopularityRanges(tracks),
      artists: [...new Set(tracks.slice(0, 10).map(t => t.artists[0]?.name).filter(Boolean))]
    };
  }

  private getPopularityRanges(tracks: SpotifyTrack[]): { year: number; count: number }[] {
    const yearCounts: { [year: number]: number } = {};
    
    tracks.forEach(track => {
      const year = new Date(track.album.release_date).getFullYear();
      if (year && year > 1950 && year < 2030) {
        yearCounts[year] = (yearCounts[year] || 0) + 1;
      }
    });

    return Object.entries(yearCounts)
      .map(([year, count]) => ({ year: parseInt(year), count }))
      .sort((a, b) => b.count - a.count)
      .slice(0, 3);
  }

  private extractGenreTerms(topTracks: SpotifyTrack[]): string[] {
    // Extract potential genre terms from track names and artists
    const genreTerms: string[] = [];
    
    for (const track of topTracks.slice(0, 10)) {
      const trackName = track.name.toLowerCase();
      const artistName = track.artists[0]?.name.toLowerCase() || "";
      
      // Common genre keywords
      const genreKeywords = [
        'pop', 'rock', 'hip hop', 'rap', 'electronic', 'edm', 'indie', 'alternative',
        'country', 'jazz', 'blues', 'classical', 'folk', 'reggae', 'funk', 'soul',
        'r&b', 'rnb', 'trap', 'house', 'techno', 'ambient', 'acoustic', 'punk'
      ];
      
      for (const keyword of genreKeywords) {
        if (trackName.includes(keyword) || artistName.includes(keyword)) {
          if (!genreTerms.includes(keyword)) {
            genreTerms.push(keyword);
          }
        }
      }
    }
    
    return genreTerms.slice(0, 5); // Return top 5 genre terms
  }
}

export const spotifyService = new SpotifyService();

File: backend/services/spotify_service.py
import spotipy
from spotipy.oauth2 import SpotifyClientCredentials, SpotifyOAuth
import os
from typing import List, Dict, Any, Optional
import asyncio
import httpx
from dotenv import load_dotenv

class SpotifyService:
    def __init__(self):
        # Load environment variables from parent directory
        load_dotenv(dotenv_path="../.env")
        self.client_id = os.getenv("EXPO_PUBLIC_SPOTIFY_CLIENT_ID")
        self.client_secret = os.getenv("SPOTIFY_CLIENT_SECRET")
        self.redirect_uri = os.getenv("SPOTIFY_REDIRECT_URI", "http://localhost:3000/callback")
        
        # Initialize Spotify client with client credentials (for public data)
        client_credentials_manager = SpotifyClientCredentials(
            client_id=self.client_id,
            client_secret=self.client_secret
        )
        self.sp = spotipy.Spotify(client_credentials_manager=client_credentials_manager)
        
        # Store user tokens (in production, this would be in a database)
        self.user_tokens = {}
    
    def get_user_spotify_client(self, access_token: str) -> spotipy.Spotify:
        """Get a Spotify client authenticated with user access token"""
        return spotipy.Spotify(auth=access_token)
    
    async def get_track_features(self, tracks: List[Dict]) -> List[Dict]:
        """Get audio features for a list of tracks"""
        track_ids = [track['id'] for track in tracks if 'id' in track]
        
        if not track_ids:
            return []
        
        try:
            # Get audio features in batches
            features = []
            batch_size = 100
            
            for i in range(0, len(track_ids), batch_size):
                batch_ids = track_ids[i:i + batch_size]
                batch_features = self.sp.audio_features(batch_ids)
                features.extend(batch_features)
            
            return features
        except Exception as e:
            print(f"Error getting track features: {e}")
            return []
    
    async def search_tracks_by_vibe(self, vibe_mode: str, limit: int = 50) -> List[Dict]:
        """Search for tracks based on vibe mode"""
        try:
            # Map vibe modes to search queries
            vibe_queries = {
                'hip-hop': 'genre:hip-hop',
                'rock': 'genre:rock',
                'pop': 'genre:pop',
                'jazz': 'genre:jazz',
                'electronic': 'genre:electronic',
                'classical': 'genre:classical',
                'country': 'genre:country',
                'r&b': 'genre:r&b',
                'reggae': 'genre:reggae',
                'blues': 'genre:blues',
                'folk': 'genre:folk',
                'indie': 'genre:indie',
                'metal': 'genre:metal',
                'punk': 'genre:punk',
                'soul': 'genre:soul',
                'funk': 'genre:funk',
                'disco': 'genre:disco',
                'house': 'genre:house',
                'techno': 'genre:techno',
                'ambient': 'genre:ambient',
                'party': 'mood:party',
                'chill': 'mood:chill',
                'energetic': 'mood:energetic',
                'romantic': 'mood:romantic',
                'sad': 'mood:sad',
                'happy': 'mood:happy',
                'workout': 'mood:workout',
                'focus': 'mood:focus',
                'sleep': 'mood:sleep'
            }
            
            query = vibe_queries.get(vibe_mode.lower(), vibe_mode)
            
            # Search for tracks
            results = self.sp.search(q=query, type='track', limit=limit)
            tracks = results['tracks']['items']
            
            # Format tracks
            formatted_tracks = []
            for track in tracks:
                formatted_track = {
                    'id': track['id'],
                    'name': track['name'],
                    'artists': track['artists'],
                    'album': track['album'],
                    'popularity': track['popularity'],
                    'preview_url': None,  # Spotify removed preview URLs from Web API
                    'external_urls': track['external_urls'],
                    'duration_ms': track['duration_ms'],
                    'explicit': track['explicit']
                }
                formatted_tracks.append(formatted_track)
            
            return formatted_tracks
            
        except Exception as e:
            print(f"Error searching tracks for vibe {vibe_mode}: {e}")
            return []
    
    async def get_track_by_id(self, track_id: str) -> Dict:
        """Get track details by ID"""
        try:
            track = self.sp.track(track_id)
            return {
                'id': track['id'],
                'name': track['name'],
                'artists': track['artists'],
                'album': track['album'],
                'popularity': track['popularity'],
                'preview_url': None,  # Spotify removed preview URLs from Web API
                'external_urls': track['external_urls'],
                'duration_ms': track['duration_ms'],
                'explicit': track['explicit']
            }
        except Exception as e:
            print(f"Error getting track {track_id}: {e}")
            return {}
    
    async def get_artist_genres(self, artist_id: str) -> List[str]:
        """Get genres for an artist"""
        try:
            artist = self.sp.artist(artist_id)
            return artist.get('genres', [])
        except Exception as e:
            print(f"Error getting artist genres for {artist_id}: {e}")
            return []
    
    async def get_user_top_tracks(self, time_range: str = "medium_term", limit: int = 50) -> List[Dict]:
        """Get user's top tracks (requires user authentication)"""
        try:
            # This would need user access token in production
            # For now, return empty list as this requires user auth
            return []
        except Exception as e:
            print(f"Error getting user top tracks: {e}")
            return []
    
    async def search_tracks(self, query: str, limit: int = 20) -> List[Dict]:
        """Search for tracks on Spotify"""
        try:
            results = self.sp.search(q=query, type='track', limit=limit)
            tracks = results['tracks']['items']
            
            formatted_tracks = []
            for track in tracks:
                formatted_track = {
                    'id': track['id'],
                    'name': track['name'],
                    'artists': track['artists'],
                    'album': track['album'],
                    'popularity': track['popularity'],
                    'preview_url': None,  # Spotify removed preview URLs from Web API
                    'external_urls': track['external_urls'],
                    'duration_ms': track['duration_ms'],
                    'explicit': track['explicit']
                }
                formatted_tracks.append(formatted_track)
            
            return formatted_tracks
            
        except Exception as e:
            print(f"Error searching tracks: {e}")
            return []
    
    async def get_recommendations_for_vibe_mode(self, vibe_mode: str, user_top_tracks: List[Dict] = None) -> List[Dict]:
        """Get tracks for a vibe mode using search (not recommendations API)"""
        try:
            all_tracks = []
            
            # Vibe-specific searches
            vibe_searches = [
                f"{vibe_mode} music",
                f"popular {vibe_mode}",
                f"{vibe_mode} hits",
                f"trending {vibe_mode}",
                f"best {vibe_mode} songs",
                f"{vibe_mode} 2024",
                f"{vibe_mode} 2023",
                f"top {vibe_mode} artists"
            ]
            
            # Search for vibe-specific tracks
            for search_term in vibe_searches:
                try:
                    tracks = await self.search_tracks(search_term, 30)
                    all_tracks.extend(tracks)
                    print(f"Found {len(tracks)} tracks for '{search_term}'")
                except Exception as e:
                    print(f"Failed to search for '{search_term}': {e}")
            
            # Add popular artists
            popular_artists = [
                "Drake", "Taylor Swift", "The Weeknd", "Billie Eilish", "Ariana Grande",
                "Ed Sheeran", "Post Malone", "Dua Lipa", "Olivia Rodrigo", "Harry Styles",
                "Bad Bunny", "Travis Scott", "Kendrick Lamar", "J. Cole", "Future"
            ]
            
            for artist in popular_artists:
                try:
                    tracks = await self.search_tracks(f"artist:{artist}", 25)
                    all_tracks.extend(tracks)
                except Exception as e:
                    print(f"Failed to search for artist '{artist}': {e}")
            
            # Remove duplicates
            unique_tracks = []
            seen_ids = set()
            for track in all_tracks:
                if track['id'] not in seen_ids:
                    unique_tracks.append(track)
                    seen_ids.add(track['id'])
            
            print(f"Found {len(unique_tracks)} unique tracks for vibe mode")
            return unique_tracks[:300]  # Return up to 300 tracks
            
        except Exception as e:
            print(f"Error getting recommendations for vibe mode: {e}")
            return []
    
    async def get_user_playlists(self) -> List[Dict]:
        """Get user's playlists (requires user authentication)"""
        try:
            # This would need user access token in production
            # For now, return empty list as this requires user auth
            return []
        except Exception as e:
            print(f"Error getting user playlists: {e}")
            return []
    
    async def get_playlist_tracks(self, playlist_id: str) -> List[Dict]:
        """Get tracks from a playlist (requires user authentication)"""
        try:
            # This would need user access token in production
            # For now, return empty list as this requires user auth
            return []
        except Exception as e:
            print(f"Error getting playlist tracks: {e}")
            return []
    
    async def like_track(self, track_id: str, access_token: str) -> bool:
        """Like a track (requires user authentication)"""
        try:
            if not access_token:
                print("No access token provided for liking track")
                return False
                
            # Use direct API call with query parameters
            headers = {
                "Authorization": f"Bearer {access_token}",
                "Content-Type": "application/json"
            }
            
            async with httpx.AsyncClient() as client:
                response = await client.put(
                    f"https://api.spotify.com/v1/me/tracks?ids={track_id}",
                    headers=headers
                )
                
                print(f"🎵 Spotify API Response - Like Track {track_id}: {response.status_code}")
                
                if response.status_code == 200:
                    print(f"✅ Successfully liked track: {track_id}")
                    return True
                elif response.status_code == 401:
                    print(f"❌ Authentication failed for track {track_id}: {response.text}")
                    return False
                else:
                    print(f"❌ Error liking track {track_id}: {response.status_code} - {response.text}")
                    return False
                    
        except Exception as e:
            if "401" in str(e) or "expired" in str(e).lower():
                print(f"Access token expired for liking track: {e}")
                return False
            print(f"Error liking track: {e}")
            return False
    
    async def unlike_track(self, track_id: str, access_token: str) -> bool:
        """Unlike a track (requires user authentication)"""
        try:
            if not access_token:
                print("No access token provided for unliking track")
                return False
                
            # Use direct API call with query parameters for DELETE
            headers = {
                "Authorization": f"Bearer {access_token}",
                "Content-Type": "application/json"
            }
            
            async with httpx.AsyncClient() as client:
                response = await client.delete(
                    f"https://api.spotify.com/v1/me/tracks?ids={track_id}",
                    headers=headers
                )
                
                print(f"🎵 Spotify API Response - Unlike Track {track_id}: {response.status_code}")
                
                if response.status_code == 200:
                    print(f"✅ Successfully unliked track: {track_id}")
                    return True
                elif response.status_code == 401:
                    print(f"❌ Authentication failed for track {track_id}: {response.text}")
                    return False
                else:
                    print(f"❌ Error unliking track {track_id}: {response.status_code} - {response.text}")
                    return False
                    
        except Exception as e:
            if "401" in str(e) or "expired" in str(e).lower():
                print(f"Access token expired for unliking track: {e}")
                return False
            print(f"Error unliking track: {e}")
            return False
    
    async def get_recommendations(self, seed_tracks: List[str], limit: int = 20) -> List[Dict]:
        """Get track recommendations based on seed tracks"""
        try:
            recommendations = self.sp.recommendations(
                seed_tracks=seed_tracks[:5],  # Max 5 seed tracks
                limit=limit
            )
            
            tracks = recommendations['tracks']
            formatted_tracks = []
            
            for track in tracks:
                formatted_track = {
                    'id': track['id'],
                    'name': track['name'],
                    'artists': track['artists'],
                    'album': track['album'],
                    'popularity': track['popularity'],
                    'preview_url': None,  # Spotify removed preview URLs from Web API
                    'external_urls': track['external_urls'],
                    'duration_ms': track['duration_ms'],
                    'explicit': track['explicit']
                }
                formatted_tracks.append(formatted_track)
            
            return formatted_tracks
            
        except Exception as e:
            print(f"Error getting recommendations: {e}")
            return []
    
    async def get_user_profile(self, access_token: str) -> Dict[str, Any]:
        """Get current user profile"""
        try:
            user_sp = self.get_user_spotify_client(access_token)
            return user_sp.current_user()
        except Exception as e:
            print(f"Error getting user profile: {e}")
            return {}
    
    async def get_user_top_tracks(self, access_token: str, time_range: str = "medium_term", limit: int = 50) -> List[Dict]:
        """Get user's top tracks"""
        try:
            user_sp = self.get_user_spotify_client(access_token)
            results = user_sp.current_user_top_tracks(time_range=time_range, limit=limit)
            return results['items']
        except Exception as e:
            print(f"Error getting user top tracks: {e}")
            return []
    
    async def get_user_top_artists(self, access_token: str, time_range: str = "medium_term", limit: int = 50) -> List[Dict]:
        """Get user's top artists"""
        try:
            user_sp = self.get_user_spotify_client(access_token)
            results = user_sp.current_user_top_artists(time_range=time_range, limit=limit)
            return results['items']
        except Exception as e:
            print(f"Error getting user top artists: {e}")
            return []
    
    async def get_user_playlists(self, access_token: str, limit: int = 50) -> List[Dict]:
        """Get user's playlists"""
        try:
            user_sp = self.get_user_spotify_client(access_token)
            results = user_sp.current_user_playlists(limit=limit)
            return results['items']
        except Exception as e:
            print(f"Error getting user playlists: {e}")
            return []
    
    async def get_playlist_tracks(self, access_token: str, playlist_id: str, limit: int = 100) -> List[Dict]:
        """Get tracks from a specific playlist"""
        try:
            user_sp = self.get_user_spotify_client(access_token)
            results = user_sp.playlist_tracks(playlist_id, limit=limit)
            return [item['track'] for item in results['items'] if item['track']]
        except Exception as e:
            print(f"Error getting playlist tracks: {e}")
            return []
    
    async def create_playlist(self, access_token: str, name: str, description: str = "", public: bool = False) -> Dict[str, Any]:
        """Create a new playlist for the user"""
        try:
            user_sp = self.get_user_spotify_client(access_token)
            user = user_sp.current_user()
            playlist = user_sp.user_playlist_create(
                user['id'], 
                name, 
                public=public, 
                description=description
            )
            return playlist
        except Exception as e:
            print(f"Error creating playlist: {e}")
            return {}
    
    async def add_tracks_to_playlist(self, access_token: str, playlist_id: str, track_uris: List[str]) -> bool:
        """Add tracks to a playlist"""
        try:
            user_sp = self.get_user_spotify_client(access_token)
            user_sp.playlist_add_items(playlist_id, track_uris)
            return True
        except Exception as e:
            print(f"Error adding tracks to playlist: {e}")
            return False
    
    async def remove_tracks_from_playlist(self, access_token: str, playlist_id: str, track_uris: List[str]) -> bool:
        """Remove tracks from a playlist"""
        try:
            user_sp = self.get_user_spotify_client(access_token)
            user_sp.playlist_remove_specific_occurrences_of_items(playlist_id, track_uris)
            return True
        except Exception as e:
            print(f"Error removing tracks from playlist: {e}")
            return False
    
    async def search_tracks(self, query: str, limit: int = 20) -> List[Dict]:
        """Search for tracks"""
        try:
            results = self.sp.search(q=query, type='track', limit=limit)
            tracks = results['tracks']['items']
            
            # Remove preview_url since Spotify removed it from Web API
            for track in tracks:
                if 'preview_url' in track:
                    track['preview_url'] = None
            
            return tracks
        except Exception as e:
            print(f"Error searching tracks: {e}")
            return []
    
    async def get_audio_features(self, track_ids: List[str]) -> List[Dict]:
        """Get audio features for tracks"""
        try:
            if not track_ids:
                return []
            
            # Process in batches of 100 (Spotify API limit)
            all_features = []
            for i in range(0, len(track_ids), 100):
                batch = track_ids[i:i + 100]
                features = self.sp.audio_features(batch)
                all_features.extend(features)
            
            return [f for f in all_features if f is not None]
        except Exception as e:
            print(f"Error getting audio features: {e}")
            return []
    
    async def get_recommendations_for_vibe_mode(self, vibe_mode: str, user_top_tracks: List[Dict], limit: int = 300) -> List[Dict]:
        """Get recommendations for a specific vibe mode using search-based approach"""
        try:
            all_tracks = []
            
            # Vibe-specific search terms
            vibe_searches = [
                f"{vibe_mode} music",
                f"popular {vibe_mode}",
                f"{vibe_mode} hits",
                f"trending {vibe_mode}",
                f"best {vibe_mode} songs",
                f"{vibe_mode} 2024",
                f"{vibe_mode} 2023",
                f"top {vibe_mode} artists"
            ]
            
            # Search for vibe-specific tracks
            for search_term in vibe_searches:
                try:
                    tracks = await self.search_tracks(search_term, 30)
                    all_tracks.extend(tracks)
                except Exception as e:
                    print(f"Error searching for '{search_term}': {e}")
            
            # Add popular tracks by searching for well-known artists
            popular_artists = [
                "Drake", "Taylor Swift", "The Weeknd", "Billie Eilish", "Ariana Grande",
                "Ed Sheeran", "Post Malone", "Dua Lipa", "Olivia Rodrigo", "Harry Styles",
                "Bad Bunny", "Travis Scott", "Kendrick Lamar", "J. Cole", "Future"
            ]
            
            for artist in popular_artists:
                try:
                    tracks = await self.search_tracks(f"artist:{artist}", 25)
                    all_tracks.extend(tracks)
                except Exception as e:
                    print(f"Error searching for artist '{artist}': {e}")
            
            # Remove duplicates by ID first
            unique_tracks = []
            seen_ids = set()
            for track in all_tracks:
                if track['id'] not in seen_ids:
                    unique_tracks.append(track)
                    seen_ids.add(track['id'])
            
            # Then remove duplicates by track name (case-insensitive)
            seen_names = set()
            final_tracks = []
            for track in unique_tracks:
                track_name = track['name'].lower().strip()
                if track_name not in seen_names:
                    seen_names.add(track_name)
                    final_tracks.append(track)
            
            print(f"🎵 Deduplication: {len(all_tracks)} -> {len(unique_tracks)} (by ID) -> {len(final_tracks)} (by name)")
            return final_tracks[:limit]
        except Exception as e:
            print(f"Error getting recommendations for vibe mode: {e}")
            return []
